Issue 63 - adding [DaisyUI tabs](https://daisyui.com/components/tab/).

mkompanets opened the pull request:
Renders tab links, radio toggle tabs and label+radio toggle tabs through the same API.

phcurado commented on lib/daisy_ui_components/tabs.ex:
```diff
@@ -0,0 +1,250 @@
+
+  use DaisyUIComponents, :component
+
+  @doc """
```
very good docs :muscle:

phcurado commented on lib/daisy_ui_components/tabs.ex:
```diff
@@ -0,0 +1,250 @@
+
+    ~H"""
+    <div class={@class} {@rest}>
+      <%= for {tab, index} <- Enum.with_index(@tab) do %>
```
I think you don't need the index from `Enum.with_index(@tab)`

```heex
<%= for tab <- @tab do %>
        <.tab_radio
          name={@name}
          title={@title} # adding the new title attribute for naming the tabs
          class={radio_tab_classes(tab)}
          aria_label={render_slot(tab)}
          active={tab[:active]}
        />
        <.tab_content :if={tab_content = render_slot(@tab)} class={tab_content_classes(tab_content)}>
          {tab_content}
        </.tab_content>
<% end %>
```

This would simplify the logic also for the content, no need for the slot `slot :tab_content` in the component attributes.

phcurado commented on lib/daisy_ui_components/tabs.ex:
```diff
@@ -0,0 +1,250 @@
+  Renders a Tabs with links as toggle by default.
+
+  ## Default Example
+  <.tabs>
```
how about we use like this:
```heex
  <.tabs>
    <:tab title="Tab 1">Tab Content</:tab>
    <:tab  title="Tab 2" active>Tab 2 Content</:tab>
    <:tab title="Tab 3" disabled />
  </.tabs>
```
This way we wouldn't need the tab content slot, since the inner_block would be the content.
`tab_link` could (for example) take the `title` attribute to add under the <a> element

mkompanets replied:
I like this. Its nice and simple, but I think the tab needs to be able to take more than just a string title. From the docs there is an example where it gets rendered with icons, so I think it should take an inner block. If we limit it to just title, icon and icon position attribute, having the tab content inside the `<:tab>` slot would simplify things quite a bit.

phcurado commented on lib/daisy_ui_components/tabs.ex:
```diff
@@ -0,0 +1,250 @@
+  ## Renders
+  <div class="tabs">
+    <label class="tab">
+      <input type="radio" name="tabs" class="tab"> Tab 1
```
From what I saw in the docs, should use aria-label as text and not the slot
```suggestion
      <input type="radio" name="tabs" class="tab"  aria-label="Tab 1" />
```

mkompanets replied:
From the docs example, the radio toggle only example puts the title into the `aria-label`. From [this example](https://daisyui.com/components/tab/#radio-tabs-lift-with-icons--tab-content) (with labels) it looks like the label wraps the radio toggle with the title rendered separate.

phcurado commented on lib/daisy_ui_components/tabs.ex:
```diff
@@ -0,0 +1,250 @@
+      assign(assigns, :class, tabs_classes(assigns))
+
+    ~H"""
+    <div class={@class} {@rest}>
```
same comment from the input radio applies here

phcurado commented on lib/daisy_ui_components/tabs.ex:
```diff
@@ -0,0 +1,250 @@
+          active={tab[:active]}
+        />
+
+        <% tab_content = Enum.at(@tab_content, index) %>
```
I think the problem with this approach is when we have tabs without content, the index would be different between the tabs and tab contents ( for example there are 3 tabs but only the last two have content).

mkompanets replied:
From my understand of how the radio input toggle works is that it toggles the sibling content element. That's why this rendering with index is a bit weird. So this assumes the first tab goes with the first content. If a content is missing, it just won't be rendered by the conditional.

phcurado replied:
Thinking on this example:

```heex
<.tabs>
    <:tab></:tab>
    <:tab active>Tab 2</:tab>
    <:tab disabled>Tab 3</:tab>
  </.tabs>
```

I think the rendered html would be wrong:

```heex
  <div role="tablist" class="tabs">
    <a role="tab" class="tab">Tab 2</a>
    <a role="tab" class="tab tab-active">Tab 3</a>
    <a role="tab" class="tab tab-disabled"></a>
  </div>
```
Becayse the first tab have no element, so the index would be mixed between the tabs no? Thats what I understood from the code

mkompanets replied:
What if we get rid of tab slot all together in favor of `<.tab>` component.

The definitions for simple link tabs could be:

```
<.tabs boder="true">
  <.tab type="link" active>
    Tab 1
    <.icon name="hero-someicon" />
  </.tab>
  <.tab type="link">
    <.icon name="hero-someicon" />
    Tab 2
  </.tab>
</.tabs>
```

For label tabs it would be:

```
<.tabs boder="true">
  <.tab type="label" active>
    Tab 1
    <.icon name="hero-someicon" />
  </.tab>
  <.tab_content>
    Show me first
  </.tab_content>
  <.tab type="link">
    <.icon name="hero-someicon" />
    Tab 2
  </.tab>
   <.tab_content>
    Show me second
  </.tab_content>
</.tabs>
```

So `<tab_content>` would be its own component, but it will be rendered wherever you place it. We won't have to figure it out so it will be defined just like in the daisy docs

phcurado replied:
this would work also, very good idea.
And since they are separated components, we can even offer slots in future (since slots can be just a wrap over these components).

phcurado reviewed (commented):
Looking very good, added a couple of comments

mkompanets commented:
@phcurado I updated the code using the approach we discussed. I still need to add a `label` type.

I added the `title` attribute which made things a bit messy than I would like, but I think it makes sense to have. The idea is that if you provide the title, you cannot provide the inner block. The inner block can just be ignored on most types except `radio`, where a title is required and I added a raise in case it isn't since inner block never makes sense.

Let me know what you think.

phcurado commented on lib/daisy_ui_components/tabs.ex:
```diff
@@ -0,0 +1,399 @@
+
+  slot :inner_block
+
+  def tab_button(%{title: _title} = assigns) do
```
I think this one is not need it right? I could not find tab button on the daisyUI examples

phcurado commented on lib/daisy_ui_components/tabs.ex:
```diff
@@ -0,0 +1,399 @@
+  attr :class, :any, default: nil
+  attr :size, :string, values: sizes()
+  attr :tabs_style, :string, values: ~w(box border lift)
+  attr :tabs_position, :string, values: ~w(top bottom)
```
what do you think we only call it `position`? Since `tabs` is implied on this component.

mkompanets replied:
Sounds good!

phcurado commented on lib/daisy_ui_components/tabs.ex:
```diff
@@ -0,0 +1,399 @@
+
+  attr :class, :any, default: nil
+  attr :size, :string, values: sizes()
+  attr :tabs_style, :string, values: ~w(box border lift)
```
I would call `border_style` or just `border` on this one

mkompanets replied:
Hmm so it could be `<.tab border_style="border">`? I don't like using the word `style` there either. Maybe `display_style`? We could also make each style a boolean value. I think that's how those classes work. So it could be `border=true` or `lift="true"` or something like that

phcurado replied:
that's a good point. I know that in daisyUI they introduced these border styles in a lot of components in the new version so would be good to have some similar attributes across components. 

I usually use the Boolean assigns when we can stack these classes, I think stacking these border elements is not possible but I really like how the API would like from your example. I would say let's do the Boolean attributes

mkompanets replied:
@phcurado I updated `tabs_position` to `position` and converted `tabs_style` to boolean. I noticed that setting `<tabs lift>` would work, but `<.tabs lift="true">` didn't. I updated `maybe_add_class` utility function to accept string "true" in addition to boolean since I think HTML can rendered with string "true" in some cases. The tests seem to all pass, but let me know if there are issues with that.

phcurado replied:
all good

phcurado reviewed (commented):
Thanks for this, looks really awesome. I added a few extra comments

phcurado reviewed (approved):
awesome :rocket:
